WordPress/src/main/java/org/wordpress/android/util/analytics/service/InstallationReferrerServiceStarter.java

```diff
@@ -0,0 +1,53 @@
+package org.wordpress.android.util.analytics.service;
+
+import android.app.job.JobInfo;
+import android.app.job.JobScheduler;
+import android.content.ComponentName;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+import android.os.PersistableBundle;
+
+import org.wordpress.android.util.AppLog;
+
+import static org.wordpress.android.JobServiceId.JOB_INSTALL_REFERRER_SERVICE_ID;
+
+public class InstallationReferrerServiceStarter {
+    public static final String ARG_REFERRER = "arg_referrer";
+    public static void startService(Context context, String referrer) {
+        if (context == null) {
+            return;
+        }
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+            Intent intent = new Intent(context, InstallationReferrerService.class);
+            if (referrer != null) {
+                intent.putExtra(ARG_REFERRER, referrer);
+            }
+            context.startService(intent);
+        } else {
+            // schedule the JobService here for API >= 26. The JobScheduler is available since API 21, but
+            // since we are only using it elsewhere on API >= 26 to not break old behavior for pre-existing services,
+            // let's stick to that version as well here.
+            ComponentName componentName = new ComponentName(context, InstallationReferrerJobService.class);
+            PersistableBundle extras = new PersistableBundle();
+            if (referrer != null) {
+                extras.putString(ARG_REFERRER, referrer);
+            }
+
+            JobInfo jobInfo = new JobInfo.Builder(JOB_INSTALL_REFERRER_SERVICE_ID, componentName)
+                    .setRequiresCharging(false)
+                    .setRequiredNetworkType(JobInfo.NETWORK_TYPE_ANY)
+                    .setOverrideDeadline(0) // if possible, try to run right away
+                    .setExtras(extras)
+                    .build();
+
+            JobScheduler jobScheduler = (JobScheduler) context.getSystemService(Context.JOB_SCHEDULER_SERVICE);
+            int resultCode = jobScheduler.schedule(jobInfo);
```

Comment:
Since `jobScheduler` can be `null`, should we add a null-check here?

Reply:
I think while ultimately possible, it can't be null as per the check for `Build.VERSION.SDK_INT` we're making a few lines above (the passed string is one that is known to be defined since the checked API level). It would fail if the passed string did not exist, but then there's a linter check that would throw a warning, because it inevitably needs to be one of the strings defined with `@ServiceName`  notation - if you look into `Context` source code the declaration goes like this:

```
public abstract @Nullable Object getSystemService(@ServiceName @NonNull String name);
```
(so, it can't be _any_ string).

If, it is after all the case that this returns `null` on an actual device (which IIUC would be a weird thing) then I think it would be preferred that we fail fast (i.e. let it crash) and be able to determine what happened analyzing afterwards, than failing silently. It might be arguable that we don't really want to halt execution when failing to track an install (it shouldn't be vital), but then all other instances of `JobScheduler jobScheduler` do the same in the codebase so it would fail shortly after anyway.

Does that make sense?